File: free_claude_code/anthropic.py
from __future__ import annotations

import json
import math
import time
import uuid
from typing import Any, Dict, List, Optional


def _text_from_blocks(content: Any) -> str:
    if isinstance(content, str):
        return content
    if not isinstance(content, list):
        return ""
    parts = []
    for b in content:
        if isinstance(b, dict) and b.get("type") == "text":
            parts.append(b.get("text", ""))
    return "\n".join(parts)


def _tool_result_content(block: Dict[str, Any]) -> str:
    c = block.get("content", "")
    if isinstance(c, str):
        return c
    if isinstance(c, list):
        out = []
        for item in c:
            if isinstance(item, str):
                out.append(item)
            elif isinstance(item, dict) and item.get("type") == "text":
                out.append(item.get("text", ""))
            else:
                out.append(json.dumps(item, ensure_ascii=False))
        return "\n".join(out)
    return json.dumps(c, ensure_ascii=False)


def anthropic_messages_to_openai(messages: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    out: List[Dict[str, Any]] = []
    for msg in messages or []:
        role = msg.get("role")
        content = msg.get("content", "")
        if isinstance(content, str):
            out.append({"role": role, "content": content})
            continue
        if not isinstance(content, list):
            out.append({"role": role, "content": ""})
            continue
        if role == "assistant":
            text = _text_from_blocks(content) or None
            calls = []
            for b in content:
                if isinstance(b, dict) and b.get("type") == "tool_use":
                    calls.append({
                        "id": b.get("id") or "call_" + uuid.uuid4().hex,
                        "type": "function",
                        "function": {"name": b.get("name"), "arguments": json.dumps(b.get("input") or {})},
                    })
            item: Dict[str, Any] = {"role": "assistant", "content": text}
            if calls:
                item["tool_calls"] = calls
            out.append(item)
            continue
        pending_text: List[str] = []
        for b in content:
            if not isinstance(b, dict):
                continue
            if b.get("type") == "text":
                pending_text.append(b.get("text", ""))
            elif b.get("type") == "tool_result":
                if pending_text:
                    out.append({"role": "user", "content": "\n".join(pending_text)})
                    pending_text = []
                out.append({"role": "tool", "tool_call_id": b.get("tool_use_id"), "content": _tool_result_content(b)})
        if pending_text:
            out.append({"role": role, "content": "\n".join(pending_text)})
    return out


def anthropic_tools_to_openai(tools: Any) -> Optional[List[Dict[str, Any]]]:
    if not isinstance(tools, list) or not tools:
        return None
    return [{
        "type": "function",
        "function": {
            "name": t.get("name"),
            "description": t.get("description", ""),
            "parameters": t.get("input_schema") or {"type": "object", "properties": {}},
        },
    } for t in tools if isinstance(t, dict) and t.get("name")]


def system_to_openai(system: Any) -> List[Dict[str, str]]:
    if not system:
        return []
    if isinstance(system, str):
        return [{"role": "system", "content": system}]
    if isinstance(system, list):
        text = "\n".join(x if isinstance(x, str) else x.get("text", "") for x in system if x)
        return [{"role": "system", "content": text}] if text else []
    return []


def build_openai_request(body: Dict[str, Any], provider_model: str, default_max_tokens: int) -> Dict[str, Any]:
    req: Dict[str, Any] = {
        "model": provider_model,
        "messages": system_to_openai(body.get("system")) + anthropic_messages_to_openai(body.get("messages") or []),
        "max_tokens": body.get("max_tokens") or default_max_tokens,
        "stream": bool(body.get("stream")),
    }
    for k in ("temperature", "top_p", "stop"):
        if k in body and body[k] is not None:
            req[k] = body[k]
    tools = anthropic_tools_to_openai(body.get("tools"))
    if tools:
        req["tools"] = tools
    tc = body.get("tool_choice")
    if isinstance(tc, dict):
        if tc.get("type") == "auto":
            req["tool_choice"] = "auto"
        elif tc.get("type") == "any":
            req["tool_choice"] = "required"
        elif tc.get("type") == "tool" and tc.get("name"):
            req["tool_choice"] = {"type": "function", "function": {"name": tc["name"]}}
    return req


def finish_to_stop_reason(finish: Optional[str]) -> str:
    if finish == "tool_calls":
        return "tool_use"
    if finish == "length":
        return "max_tokens"
    if finish == "content_filter":
        return "stop_sequence"
    return "end_turn"


def openai_to_anthropic(data: Dict[str, Any], request_model: str) -> Dict[str, Any]:
    choice = (data.get("choices") or [{}])[0]
    msg = choice.get("message") or {}
    content: List[Dict[str, Any]] = []
    if msg.get("content"):
        content.append({"type": "text", "text": msg.get("content")})
    for tc in msg.get("tool_calls") or []:
        args = tc.get("function", {}).get("arguments") or "{}"
        try:
            parsed = json.loads(args)
        except Exception:
            parsed = {}
        content.append({
            "type": "tool_use",
            "id": tc.get("id") or "toolu_" + uuid.uuid4().hex,
            "name": tc.get("function", {}).get("name"),
            "input": parsed,
        })
    if not content:
        content = [{"type": "text", "text": ""}]
    usage = data.get("usage") or {}
    return {
        "id": "msg_" + uuid.uuid4().hex,
        "type": "message",
        "role": "assistant",
        "model": request_model or "free-claude-code",
        "content": content,
        "stop_reason": finish_to_stop_reason(choice.get("finish_reason")),
        "stop_sequence": None,
        "usage": {"input_tokens": usage.get("prompt_tokens", 0), "output_tokens": usage.get("completion_tokens", 0)},
    }


def estimate_tokens(obj: Any) -> int:
    text = json.dumps(obj, ensure_ascii=False)
    return max(1, math.ceil(len(text) / 4))

File: my_claudecode_go/anthropic.go
package main

import (
	"encoding/json"
	"fmt"
	"strings"
)

type AnthropicMessage struct {
	Role    string `json:"role"`
	Content any    `json:"content"`
}

type AnthropicRequest struct {
	Model       string             `json:"model"`
	Messages    []AnthropicMessage `json:"messages"`
	System      any                `json:"system"`
	Stream      bool               `json:"stream"`
	MaxTokens   int                `json:"max_tokens"`
	StopSeqs    []string           `json:"stop_sequences,omitempty"`
	Temperature *float64           `json:"temperature,omitempty"`
	Tools       []any              `json:"tools,omitempty"`
	ToolChoice  any                `json:"tool_choice,omitempty"`
}

type OpenAIMessage struct {
	Role    string `json:"role"`
	Content any    `json:"content"`
	Name    string `json:"name,omitempty"`
}

type OpenAIRequest struct {
	Model       string          `json:"model"`
	Messages    []OpenAIMessage `json:"messages"`
	Stream      bool            `json:"stream"`
	MaxTokens   int             `json:"max_tokens"`
	Temperature *float64        `json:"temperature,omitempty"`
	Tools       []any           `json:"tools,omitempty"`
	ToolChoice  any             `json:"tool_choice,omitempty"`
}

func BuildOpenAIRequest(anthropic AnthropicRequest, modelID string, maxTokens int) OpenAIRequest {
	req := OpenAIRequest{
		Model:       modelID,
		Stream:      anthropic.Stream,
		MaxTokens:   anthropic.MaxTokens,
		Temperature: anthropic.Temperature,
		Tools:       anthropic.Tools,
		ToolChoice:  anthropic.ToolChoice,
	}
	if req.MaxTokens == 0 {
		req.MaxTokens = maxTokens
	}

	messages := []OpenAIMessage{}
	if anthropic.System != nil {
		systemText := ""
		switch v := anthropic.System.(type) {
		case string:
			systemText = v
		case []any:
			for _, part := range v {
				if m, ok := part.(map[string]any); ok && m["type"] == "text" {
					systemText += fmt.Sprint(m["text"])
				}
			}
		}
		if systemText != "" {
			messages = append(messages, OpenAIMessage{Role: "system", Content: systemText})
		}
	}

	for _, msg := range anthropic.Messages {
		messages = append(messages, OpenAIMessage{Role: msg.Role, Content: msg.Content})
	}
	req.Messages = messages
	return req
}

func EstimateTokens(anthropic AnthropicRequest) int {
	// Crude estimation
	count := 0
	if s, ok := anthropic.System.(string); ok {
		count += len(strings.Fields(s))
	}
	for _, m := range anthropic.Messages {
		if s, ok := m.Content.(string); ok {
			count += len(strings.Fields(s))
		}
	}
	return int(float64(count) * 1.3)
}

func OpenAIToAnthropic(openAIResponse map[string]any, modelID string) map[string]any {
	choices, _ := openAIResponse["choices"].([]any)
	if len(choices) == 0 {
		return map[string]any{"error": "no choices in response"}
	}
	choice := choices[0].(map[string]any)
	message := choice["message"].(map[string]any)

	content := []any{}
	if text, ok := message["content"].(string); ok && text != "" {
		content = append(content, map[string]any{"type": "text", "text": text})
	}

	toolCalls, _ := message["tool_calls"].([]any)
	for _, tc := range toolCalls {
		tcm := tc.(map[string]any)
		fn := tcm["function"].(map[string]any)
		var args map[string]any
		json.Unmarshal([]byte(fmt.Sprint(fn["arguments"])), &args)
		content = append(content, map[string]any{
			"type": "tool_use",
			"id":   tcm["id"],
			"name": fn["name"],
			"input": args,
		})
	}

	stopReason := "end_turn"
	if choice["finish_reason"] == "tool_calls" {
		stopReason = "tool_use"
	}

	return map[string]any{
		"id":           "msg_go_" + fmt.Sprint(openAIResponse["id"]),
		"type":         "message",
		"role":         "assistant",
		"model":        modelID,
		"content":      content,
		"stop_reason":  stopReason,
		"stop_sequence": nil,
		"usage": map[string]any{
			"input_tokens":  openAIResponse["usage"].(map[string]any)["prompt_tokens"],
			"output_tokens": openAIResponse["usage"].(map[string]any)["completion_tokens"],
		},
	}
}
